x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java

```diff
@@ -620,22 +620,26 @@ public PlanFactory visitJoinCommand(EsqlBaseParser.JoinCommandContext ctx) {
         }
 
         return p -> {
-            p.forEachUp(UnresolvedRelation.class, r -> {
-                for (var leftPattern : Strings.splitStringByCommaToArray(r.indexPattern().indexPattern())) {
-                    if (RemoteClusterAware.isRemoteIndexName(leftPattern)) {
-                        throw new ParsingException(
-                            source(target),
-                            "invalid index pattern [{}], remote clusters are not supported in LOOKUP JOIN",
-                            r.indexPattern().indexPattern()
-                        );
-                    }
-                }
-            });
-
+            checkForRemoteClusters(p, source(target), "LOOKUP JOIN");
             return new LookupJoin(source, p, right, joinFields);
         };
     }
 
+    private void checkForRemoteClusters(LogicalPlan plan, Source source, String commandName) {
```

Comment:
It feels a bit weird that we're running the same code 3 times from 3 different places just to get different error message... I wonder if there isn't a way to somehow capture it better so we don't need to walk the whole (sub-)tree each time anew.

Reply:
this is just reusing the same approach we had for LOOKUP JOIN.
even on main right now, if an ES|QL query has 3 LOOKUP JOIN commands, we are doing the same check 3 times.
I thought that if we had a better way to do it this would have been flagged in https://github.com/elastic/elasticsearch/pull/120277
also it's just a temporary limitation, since FORK/COMPLETION/FORK are all going to be tech preview features first and CCS support will be done as part of the transition to GA.

Reply:
it looks like we also have another place in `LogicalPlanBuilder` where walk the plan looking for `UnresolvedRelation`:

https://github.com/elastic/elasticsearch/blob/4e926ae41a9d33332bfc9e117e11abbeae13ee60/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java#L316-L325

just wanted to check that it's a pattern we use in some other places as well, not just for CCS check 🤔

Reply:
Yeah we do such things repeatedly and I am not a big fan of this though I am not sure yet how to fix it. It will also become more complicated with CPS since we may not be able to know whether certain index pattern means remote or local without wider context, which is not available at that point. Not sure what to do with it, maybe these checks can be moved to a different place or checked in some other way. Needs more thinking. I guess it's ok for now.